### ISSUE TITLE
[Enhancement] ttrss labels feed

### ISSUE THREAD

**Issue by pcause** [newsboat/newsboat#210]
It would be quite helpful when you enumerate the ttrss feeds to also get the labels feed so I can view the items from that feed.  I use ttrss filters to identify various items of interest and don't want to have to duplicate these all in newsboat. 

if you do, when you read a feed's items, it any of the items have a ttrss label, how about including the label header in the content displayed.  then I can at least search for the labels.  

and if you don't want to fetch the label feed can you at least add a header item to the article content with the ttrss labels?

thanks in advance

**Comment by Minoru** [comment]
TT-RSS labels are converted into Newsboat tags—press `t` to open the dialog listing them all. Is that enough for your needs?

**Comment by pcause** [comment]
no, ttrss feed categories are translated into newsboat labels.  ttrss labels are different.  think of them as a tag cloud.  there is no good equivalent in newsboat since an article can have multiple labels so just putting them in a header at the top of the article content would be enough so that one could create a newsboat filter that has the ttrss tag and then see all articles that have that ttrss label.

**Comment by Minoru** [comment]
Ah, sorry. Every service has some kind of categorization, but everybody calls it differently. I keep forgetting that TT-RSS has both tag-like and folder-like categorization (labels and categories, respectively).

I don't know how to fit this with things that already exist in Newsboat. We already have flags, which are per-article marks (just like TT-RSS labels), but flags are single English letters, not strings. Extending them to strings is quite a chunk of work.

Adding something just for TT-RSS is out of the question. Newsboat is not a client *for* any of its remote APIs; it's a program that uses existing services to fetch feeds.

I don't think adding labels to the contents of the article would be a good idea, because messing with people's data is wrong. If labels are to be added, they should be in the header—but see my points above.

That leaves us with the "labels feed" approach, which I can't figure out. I read [the docs](https://git.tt-rss.org/fox/tt-rss/wiki/GeneratedFeeds) and [some thread on the forum](https://tt-rss.org/oldforum/viewtopic.php?t=2467), but I don't have a TT-RSS instance to try this on. My questions are:
1. looks like TT-RSS can return both per-label feeds (if we specify id < -10) and some "label feed" (is_cat=1 and id=-2). Which one do you mean?
2. a follow-up to the previous point: what does "label feed" contain?
3. how do we discover which label IDs are available?

**Comment by Minoru** [comment]
@noctux, do I misunderstand this issue? Do TT-RSS labels fit with anything that already exists in Newsboat?

**Comment by pcause** [comment]
There are feeds that show the tt-rss items that have a label assigned.  As indicated about you can get the list of these in the same way you get a feed list, the names of the feeds are the labels so when you add the feed name as a newsboat label you've got something we can use the tag command on to see the labels and items.   All of the code to get this would be in the ttrss specific code.

The only wrinkle is that an item in a label feed is also an item in some other feed.  That means that if you fetch all feeds, read and mark read using the label feed and then head to another feed with the same item *without* reloading that other feed the item will appear as still unread.  Minor issue issue.

BTW, in the docs there doesn't seem to be an option to reload a feed on selecting a feed to view items.  This would fix the minor issue and be useful on its own.

**Comment by Minoru** [comment]
> … so when you add the feed name as a newsboat label …

That's the point where you lost me. Can you explain in other words, maybe?

I think you answered the first of my questions. Label feeds are indeed per-label, i.e. each label feed contains items that have a specified label. I'm still not clear if you meant these, since you wrote "label*s* feed" in your original post, and TT-RSS appear to have something with that name, too.

The third question I take back. I re-read the docs and now see that to discover label IDs, all we need to do is call `getFeeds` API method with `cat_id=-2`.

> The only wrinkle is that an item in a label feed is also an item in some other feed.

Newsboat actually has a big problem with this, see akrennmair/newsbeuter#193.

> BTW, in the docs there doesn't seem to be an option to reload a feed on selecting a feed to view items.

JFTR: I believe this was reported separately as #211.

Thank you for being so patient with explanations, @pcause; it's appreciated!

**Comment by pcause** [comment]
First, thanks for taking the time to engage and understand.

When newsboat gets the feed list from ttrss it creates a newsboat tag for the feed that is the ttrss category.  So if I have a category called greatsoftware and udner that there is a feed for newsboat-releases, when I hit t on the main newsboat screen I see a newsboat tag names greatsoftware and if I select that I see newsboat-releases feed.

In ttrss, "labels" is a category, and within that category there are feeds each of whose name is the name of the ttrss label.  So if I have a ttrss label called feedreaders, which I apply to article about rss news readers,  then the "Labels"  category has a feed called feedreader.  When newsboat gets the feed under "Labels" it would create a newsboat tag called Labels.  When I hit t, i'll see the categories I now see plus this new one, Labels.  When I select Labels the names of the feeds I see are the names of the ttrss labels, and so when I select one I see all the articles I've applied that ttrss label to.

Basically everything in newboat outside of the ttrss specific code just works if you handle getting the Labels category, which is a virtual category where the feeds in the category are virtual feeds.

The issue you mention, #193, is a different problem.  ttrss labels don't introduce new feeds with the same name (newsboat label).  What adding labels will introduce is that two different feeds (the one for the ttrss label and the one that the content originally came from) will have an article with the same URL.  The first question is when fetching articles for a feed, will getting a second article URL that matches an article URL you already have  cause a problem.  Looking at the cache.cpp code where the tables for the cache are created, it doesn't look like it will, as the item URL isn't specified to be unique and an item URL is always coupled with a feed url in the same row.  So feedurl+itemurl will be different for the two. items  If I am right, then the next question is whether marking an article read in one place will mark it read universally, because to mark read you do a loo up on the URL.   Again, looking at the ttrss.cpp and itemview_formaction.cpp it appears you use a guid to fetch items from the list displayed, so newsboat will use this to find the data for the item and then uses that data to call the ttrss specific code.  That means that newsboat will  just mark a single item for the feed read, and leave the other duplicate item in its unread state.  We'll have the article appear agin when you open the second feed even though its been really marked it read in one feed it is still unread in the other.  Of course, when you send the state change to ttrss, it knows there is only one article and so if the second feed is reloaded, the article will appear unread.  In this case my suggestion for auto-reloading a feed solves the problem of seeing the entry again as unread, because when you enter the other feed reloading will say the article is read.  But even without the auto-reload the problem is minor.

Hope this helps.  Seems to me all that is needed is to add the labels category and its feeds and all will work, even without my autoload enhancement suggestion.

**Comment by Minoru** [comment]
Thanks! I finally understand how label feeds and labels category relate. They should fit with Newsboat well.

I'm not sure, though, if label feeds should be always present. I believe Newsboat can't omit them from feedlist without omitting them from taglist, but if they're always in feedlist, that might annoy some users. That suggests adding an option to enable them, which would be off by default. But I'll be happier not to add another option if that can be helped.

I've edited the issue link in your post—it should point to Newsbeuter's issue tracker, not this one. The title of that issue is misleading, because the problem turned out to be caused by item GUIDs. In essence, if two feeds contain an item with the same GUID, that item will appear in *one* of the feeds, depending on the order in which feeds are updated.

So I guess this issue is blocked on akrennmair/newsbeuter#193 now? I started working on 193 a while ago, but it touches upon the cache layer that underlies the whole program, and I quickly broke stuff when I changed anything. That's why I switched to test coverage and refactoring this year, but sadly, the effort progresses very slowly.

**Comment by pcause** [comment]
actually, I think ttrss labels feed is always there it may just not have ave feeds under it.  Remember it is a ttrss category and you'll turn that into a label for a newsboat feed only if the ttrss labels category has feeds under it that you then tag,

The issue seemed to be about feed urls, but you are the expert.   i'll look at the tt-rss code and see if the id returned by the api call for the item when it is from one of the ttrss labels feeds is the same as returned by the regular feed.  if so, i guess same issue.

**Comment by Minoru** [comment]
> The user has to manually tag those articles

Judging by the docs, TT-RSS has filters that [let users auto-assign labels](https://git.tt-rss.org/fox/tt-rss/wiki/ContentFilters#matching-articles-and-applying-actions).

I'm not sure I understood what you're saying, though. Are you saying that Newsboat users should be able to add a label to an article from within Newsboat?

> easily emulate label-feeds with query-feeds, easing implementation, and making it even more versatile (e.g., by combining them with time ranges or other filter expressions).

Yay, that sounds much more Newsboat-like! Though users might not be very happy to manually create query feeds for each of their labels. I actually had simiilar same gripe with Mutt: before I learned about `imap_check_subscribed`, I used to manually add mailboxes that corresponded to GMail labels. It was annoying, because labels are created in GMail web-interface, and I always forgot to update my muttrc.

In summary, we have *two* solutions now:
* add label feeds (one feed per TT-RSS label), possibly hidden behind an option. This will only impact TT-RSS users. Blocked on akrennmair/newsbeuter#193;
* expose new attribute for articles, `labels`, which would be populated either from TT-RSS labels, or from corresponding tag in Atom/RSS.

The best thing is, they aren't mutually exclusive, and can be implemented in any order. Patches welcome!

**Comment by noctux** [comment]
> Judging by the docs, TT-RSS has filters that let users auto-assign labels.
> I'm not sure I understood what you're saying, though. Are you saying that Newsboat users should be able to add a label to an article from within Newsboat?

I do not really have any stake in that game as I do not actively use labels personally. However, I just wanted to note that a way to add labels might be desireable to make this feature really usable to some users. (Imagine an overly eager automatic filter expression, where one wants remove some of the automatic labels. However, this is once more something that would be much more reliable given a proper two-way sync model, as proposed in #16).

> Though users might not be very happy to manually create query feeds for each of their labels.

This is something that should be fairly easy to implement once the general query-feed-for-label concept is there, e.g., by adding an option to the config file and using `getLabels` from the upstream API to actually automatically create those feeds on startup, like we do for the normal feeds as well.

> In summary, we have two solutions now:

I'm biased there, but I would personally go with 2), especially since the TTRSS sync is still a bit slow, and adding a lot of duplicates will make the situation even worse. But anyways, yeah, patches are always welcome :)

As for 2), it might be worthwhile to combine that one with #162, which touches the query-syntax anyways and might be just a straightforward extension from there.

**Comment by pcause** [comment]
> In summary, we have two solutions now:

    add label feeds (one feed per TT-RSS label), possibly hidden behind an option. This will only impact TT-RSS users. Blocked on akrennmair/newsbeuter#193;

    expose new attribute for articles, labels, which would be populated either from TT-RSS labels, or from corresponding tag in Atom/RSS.

The best thing is, they aren't mutually exclusive, and can be implemented in any order. Patches welcome!

For a tt-rss user the first would be the most convenient.  tt-rss does a really good job with their filters and labels and you can even add plugins to enhance this.  a rich system.

A comment on the second option..  I find the tags in an RSS feed article less than useful, which is why I use a tt-rss filter to create my labels.  I look for specific words/phrases not to match the topics and projects I have on my plate which is not a match for the content creator's view of where the article fits and how they tag it..  I want to content filtered and organized for reading to get me the most relevant articles to my projects collected and presented to match my ongoing projects.

As you point out the two things aren't mutually exclusive.  I'll keep an eye out for the caching bug to be fixed.

But thanks for all the effort to understand this topic.

**Comment by Minoru** [comment]
@noctux,

> As for 2), it might be worthwhile to combine that one with #162, which touches the query-syntax anyways and might be just a straightforward extension from there.

Changing the parser is only one of multiple steps required to add a new attribute, and probably the smallest one—adding a new attribute to matchers and updating the documentation will surely outweigh it. So I don't consider those two issues related.

@pcause,
> A comment on the second option.. I find the tags in an RSS feed article less than useful, which is why I use a tt-rss filter to create my labels.

Users can ignore `labels` attribute if they want, or they can take it into account in their query feeds. The point here is to give "vanilla Newsboat" users the same power as "TT-RSS Newsboat" users.

But you got me thinking about two other things:
1. what if sometime in the future, we decide to expose RSS/Atom tags fetched by TT-RSS? Where would we put them? `labels` would be already taken by TT-RSS labels.
2. say a TT-RSS user writes s couple Newsboat query feeds to further filter their items. Then they decide to use Newsboat *without* TT-RSS. Their query feeds will immediately break, because `labels` would be populated by RSS/Atom tags now, not hand-crafted TT-RSS labels.

That calls for `labels` to be renamed to `ttrss-labels`, and a separate attribute for RSS `<category>` tags.

**Comment by pcause** [comment]
I just spent more time reading the cache code.  I don't think that #193 is going to be an issue.  For a ttrss labels feed the feed URL is different than the feed URL for the base article containing the article.  The article link will be the same between the article appearing in the ttrss labels feed and the original feed, but looking at cache newsboat doesn't use that url for any sql queries.  it seem items are fetched, updated, etc on the item guid.    and evne when you have a feed url/guid pair for an item that has a label these will be different for one will have the feed URL for the ttrss label feed and the other for the feed that originally contained the article.

So it seems to me it should be pretty safe to add the ttrss labels feeds to newsboat within the ttrss code and it should break anything outside of that code.  Is there some other cache code outside of cache.cpp that I should look at.

**Comment by Minoru** [comment]
Once again: it's "akrennmair/newsbeuter#193", not just "#193". Those two are different issues.

Newsboat considers feed URL part of an item's data, which is stored in `rss_item` table. When feed `A` adds an item with GUID `G`, that item is stored in the table. When another feed, `B`, adds an item with the same GUID, it overwrites the entry in the table. From that moment on, the item belongs to feed `B`. It's no longer part of feed `A`.

In other words:
1. items are distinguished by their GUIDs;
2. any given item only belongs to one feed.

This means that akrennmair/newsbeuter#193 would indeed apply to TT-RSS label feeds.

**Comment by pcause** [comment]
I think there is an easy work around for this.  When you get the items from an ttrss feed instead of using the "id" field of the item retuned by ttrss as the guid, you could use the ttrss feed id plus the returned item id.   The feed is in ttrss is a number.  tt-rss already returns a URI as the "link" field for every item.   Looking at the newsboat rss parser, it seems to set the item link to the the guid only if the feed doesn't give you a link and only if the guid has the permalink attribute.  So somehting like I sugggest for the ttrss api code could work.   This is a small change to the tt-rss api code, as setting the guid is now 2 lines of code the the feed id is already passed into the  code that sets the guid.  

```
int id = item_obj["id"];
item.guid = strprintf::fmt("%d", id);
```

id is a bad variable name as this is also the name of the feed id argument to the function.  so rename that variable and then build the guid from the feed id and the item id.

**Comment by noctux** [comment]
However, that way, both items will not share a guid and therefore, there is no sync of the read state of the given article, leading to a large number of duplicates. You mentioned that before and suggested autoreloading at feed-switch as a possible mitigation. To me, this sound like a big hack to support a rather particular usecase. Is there any chance for a better soluton until akrennmair/newsbeuter#193 is fixed?

**Comment by Minoru** [comment]
@pcause, imagine a user who already has a database populated by some items. When they upgrade to Newsboat with the workaround, all their feeds would become full of unread items, because all item GUIDs changed.

When we finally fix akrennmair/newsboat#193, we'll have to remove the workaround, and that will make everything unread *again*, because GUIDs changed.

And we can't avoid removing the workaround because it conflicts with sharing state between items. Identical items *should* have identical GUIDs. You can read more about that idea in akrennmair/newsbeuter#249.
